`DiagnosticSource` Filtering Behaviour

While digging into the implementation details and advanced scenarios for `DiagnosticSource`, I ran into something slightly surprising that was not covered by the docs. If you subscribe multiple observers to the same notification source, those subscribing with a filter may still receive notifications from `OnNext` if any other subscribers require the notification.

**Repro:**
```c#
using System.Diagnostics;
using Test;

DiagnosticListener.AllListeners.Subscribe(new DiagnosticListenersObserver());

DoStuff.DoIt();

namespace Test
{
    public class DoStuff
    {
        internal static readonly DiagnosticListener Listener = new("MySource");

        public static void DoIt()
        {
            if (Listener.IsEnabled("DoIt"))
                Listener.Write("DoIt", null);
        }
    }

    public class DiagnosticListenersObserver : IObserver<DiagnosticListener>
    {
        public void OnCompleted() { }
        public void OnError(Exception error) { }

        public void OnNext(DiagnosticListener value)
        {
            if (value.Name == "MySource")
            {
                value.Subscribe(new Observer("Unfiltered"));
                value.Subscribe(new Observer("Filter everything"), n => false);
            }
            else
                Console.WriteLine(value.Name);
        }
    }

    public class Observer(string name) : IObserver<KeyValuePair<string, object?>>
    {
        public string Name { get; } = name;

        public void OnCompleted() { }
        public void OnError(Exception error) { }

        public void OnNext(KeyValuePair<string, object?> value) => 
            Console.WriteLine($"{Name}: {value.Key}");
    }
}
```

**Console output:**
```
Filter everything: DoIt
Unfiltered: DoIt
```

As far as I can see, this is not explicitly called out in the docs, so I wanted to validate this is expected. Presumably, observers subscribed with a filter must recheck their filtering to handle only intended notifications in `IObserver<KeyValuePair<string, object?>>.OnNext`. If these assumptions are correct, I can raise a docs issue for this and perhaps help with a PR to add a note to the docs. 

On a related point, I have a query about the guidance in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Diagnostics.DiagnosticSource/src/DiagnosticSourceUsersGuide.md

Under the payloads section, it states

> Note that this type needs to be made public (since the listener needs to see it) and should be under the namespace System.Diagnostics.DiagnosticSource.PayloadTypes.

If this advice is followed, however, when accessing the payload type in an observer, we would require a using directive to that namespace, which causes compiler warning CS0435, due to the namespace conflicting with the `DiagnosticSource` type. Do you think this recommendation should be amended or removed?

--- a/docs/core/diagnostics/diagnosticsource-diagnosticlistener.md
+++ b/docs/core/diagnostics/diagnosticsource-diagnosticlistener.md
@@ -212,6 +212,22 @@ The `listener.Subscribe()` call in the previous example can be replaced with the
 This efficiently subscribes to only the 'RequestStart' events. All other events will cause the `DiagnosticSource.IsEnabled()`
 method to return `false` and thus be efficiently filtered out.
 
+> [!NOTE]
+> Filtering is only designed as a performance optimization. It is possible for a listener to receive events even when they
+> do not satisfy the filter. This could occur because some other listener has subscribed to the event or because the source
+> of the event didn't check IsEnabled() prior to sending it. If you want to be certain that a given event satisfies the filter
+> you will need to check it inside the callback. For example:
+
+```C#
+    Action<KeyValuePair<string, object>> callback = (KeyValuePair<string, object> evnt) =>
+        {
+            if(predicate(evnt.Key)) // only print out events that satisfy our filter
+            {
+                Console.WriteLine("From Listener {0} Received Event {1} with payload {2}", networkListener.Name, evnt.Key, evnt.Value.ToString());
+            }
+        };
+```
+
 ##### Context-based filtering
 
 Some scenarios require advanced filtering based on extended context.

